### DISCUSSION TITLE
Invalidate keepAlive dependencies

### DISCUSSION THREAD

**Discussion by ynlvko** [rrousselGit/riverpod#2505] [Q&A]
Is it possible to invalidate (or refresh) the target provider and its dependencies?

I have something like this:
```dart
@Riverpod(keepAlive: true)
Future<Dto> getDto(GetDtoRef ref, {required int id}) {
  return networkClient.get(id);
}

@Riverpod(keepAlive: true)
Future<UiModel> getUiModel(GetUiModelRef ref, {required int id}) {
  return await ref.watch(networkClientProvider(id: id).future).toUiModel();
}
```
and when I call `ref.invalidate(getUiModelProvider(id: id))` it invalidates only `getUiModelProvider`, but `getDtoProvider` is not invalidated.

**Comment by rrousselGit** [comment]
Do the opposite instead:

Invalidate the dependency and it'll invalidate the provider too.

**Reply by ynlvko** [reply]
But I don't want the implementation details to leak into the UI. I plan to change the way I get the Dto, and I don't want to change the widget in this case.

**Reply by rrousselGit** [reply]
Make your provider a notifier then, and have it expose a custom invalidate method:

```dart

@Riverpod(keepAlive: true)
class GetUiModel extends _$GetUiModel {
  ...

  void invalidate() => ref.invalidate(getDtoProvider);
}

...

ref.read(getUiModelProvider.notifier).invalidate();
```

**Reply by ynlvko** [reply]
What do you think about `ref.onInvalidate(VoidCallback)`?

**Reply by rrousselGit** [reply]
That's something I'm against, as there are way too many thing which can trigger an invalidate.
